src/pe_crypto.cpp

```diff
@@ -101,85 +103,99 @@ namespace libtorrent
 		m_xor_mask = h.final();
 	}
 
-	int encryption_handler::encrypt(std::vector<boost::asio::mutable_buffer>& iovec)
+	std::tuple<int, std::vector<boost::asio::const_buffer>>
+	encryption_handler::encrypt(
+		aux::array_view<boost::asio::mutable_buffer> iovec)
 	{
+		using namespace boost::asio;
+
 		TORRENT_ASSERT(!m_send_barriers.empty());
 		TORRENT_ASSERT(m_send_barriers.front().enc_handler);
 
 		int to_process = m_send_barriers.front().next;
 
+		boost::asio::mutable_buffer* bufs;
+		int num_bufs;
+		bool need_destruct = false;
 		if (to_process != INT_MAX)
 		{
-			for (std::vector<boost::asio::mutable_buffer>::iterator i = iovec.begin();
-				to_process >= 0; ++i)
+			bufs = TORRENT_ALLOCA(mutable_buffer, iovec.size());
```

Comment:
correct me if I'm wrong, but I believe we only get here if are asked to process fewer bytes than everything. In that case, the iovec we pass on to the underlying crypto-plugin is likely to be different from the one passed in. we copy it into a stack allocated array here. note that boost::asio::mutable_buffer is not a pod, those objects actually have to be constructed in place below, and destructed later.